internal/configs/test_file.go

```diff
@@ -0,0 +1,341 @@
+			}
+		case "variables":
+			if tf.Variables != nil {
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Multiple \"variables\" blocks",
+					Detail:   fmt.Sprintf("This test file already has a variables block defined at %s.", tf.VariablesDeclRange),
+					Subject:  block.DefRange.Ptr(),
+				})
+				continue
+			}
+
+			tf.Variables = make(map[string]hcl.Expression)
+			tf.VariablesDeclRange = block.DefRange
+
+			vars, varsDiags := block.Body.JustAttributes()
+			diags = append(diags, varsDiags...)
+			for _, v := range vars {
+				tf.Variables[v.Name] = v.Expr
+			}
+		}
+	}
+
+	return &tf, diags
+}
+
+func decodeTestRunBlock(block *hcl.Block) (*TestRun, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+
+	content, contentDiags := block.Body.Content(testRunBlockSchema)
+	diags = append(diags, contentDiags...)
+
+	r := TestRun{
+		Name:          block.Labels[0],
+		NameDeclRange: block.LabelRanges[0],
+		DeclRange:     block.DefRange,
+	}
+	for _, block := range content.Blocks {
+		switch block.Type {
+		case "assert":
+			cr, crDiags := decodeCheckRuleBlock(block, false)
+			diags = append(diags, crDiags...)
+			if !crDiags.HasErrors() {
+				r.CheckRules = append(r.CheckRules, cr)
+			}
+		case "plan_options":
+			if r.Options != nil {
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Multiple \"plan_options\" blocks",
+					Detail:   fmt.Sprintf("This run block already has a plan_options block defined at %s.", r.Options.DeclRange),
+					Subject:  block.DefRange.Ptr(),
+				})
+				continue
+			}
+
+			opts, optsDiags := decodeTestRunOptionsBlock(block)
+			diags = append(diags, optsDiags...)
+			if !optsDiags.HasErrors() {
+				r.Options = opts
+			}
+		case "variables":
+			if r.Variables != nil {
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Multiple \"variables\" blocks",
+					Detail:   fmt.Sprintf("This run block already has a variables block defined at %s.", r.VariablesDeclRange),
+					Subject:  block.DefRange.Ptr(),
+				})
+				continue
+			}
+
+			r.Variables = make(map[string]hcl.Expression)
+			r.VariablesDeclRange = block.DefRange
+
+			vars, varsDiags := block.Body.JustAttributes()
+			diags = append(diags, varsDiags...)
+			for _, v := range vars {
+				r.Variables[v.Name] = v.Expr
+			}
+
+		}
+	}
+
+	if r.Variables == nil {
+		// There is no distinction between a nil map of variables or an empty
+		// map, but we can avoid any potential nil pointer exceptions by just
+		// creating an empty map.
+		r.Variables = make(map[string]hcl.Expression)
+	}
+
+	if r.Options == nil {
+		// Create an options with default values if the user didn't specify
+		// anything.
+		r.Options = &TestRunOptions{
+			Mode:    NormalTestMode,
+			Refresh: true,
+		}
+	}
+
+	if attr, exists := content.Attributes["command"]; exists {
+		expr, exprDiags := shimTraversalInString(attr.Expr, true)
```

Comment:
Why are we using `shimTraversalInString` here (and for "mode") instead of using the expression directly? I thought this was just for legacy syntax compatibility.

Reply:
I liked the functionality of shimTraversalInString in that it allowed users to also specify the keyword with quotation marks without crashing, so I reused it. I didn't realise it was just for older parts of the config, and have now read the doc comment more thoroughly.

I've removed referencing to it and I'm just using the expression directly now.